[JSC] Make C++ -> JS calls faster

Constellation opened the pull request:
#### fe0d0e1818d0267e38495a3935584229e1625525
<pre>
[JSC] Make C++ -&gt; JS calls faster
<a href="https://bugs.webkit.org/show_bug.cgi?id=252961">https://bugs.webkit.org/show_bug.cgi?id=252961</a>
rdar://105943509

Reviewed by Mark Lam.

1. VMEntryScope&apos;s fast path should be inlined. Do it to make it work when it is used outside of JSC framework (WebCore etc.).
2. Move VMEntryScope&apos;s didPopListener to VM so that it makes VMEntryScope&apos;s destructor super simple (purge Vector destructor code bloat).
3. Remove isCollectorBusyOnCurrentThread check. It is extremely costly for this function (5% is attributed to this!). It is too costly
   for runtime check. And we instead have ASSERT already, which should catch an issue from fuzzers (And so far, we are not seeing this).
4. RETURN_IF_EXCEPTION inside DeferTraps always get the slow path function call! We call `vm.traps().needHandling(VMTraps::AllEvents)` manually
   to avoid this.

This patch makes C++ -&gt; JS calls 8% faster.

                               ToT                     Patched

    cpp-to-js-call       85.4401+-0.4316     ^     78.9938+-0.3521        ^ definitely 1.0816x faster

* JSTests/microbenchmarks/cpp-to-js-call.js: Added.
(test):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/debugger/Debugger.cpp:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
* Source/JavaScriptCore/interpreter/CallFrame.cpp:
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::executeProgram):
(JSC::Interpreter::executeCallImpl):
(JSC::Interpreter::executeConstruct):
(JSC::Interpreter::prepareForRepeatCall):
(JSC::Interpreter::executeEval):
(JSC::Interpreter::executeModuleProgram):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::whenIdle):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::invokeDidPopListeners):
* Source/JavaScriptCore/runtime/VMEntryScope.cpp:
(JSC::VMEntryScope::setUpSlow):
(JSC::VMEntryScope::tearDownSlow):
(JSC::VMEntryScope::VMEntryScope): Deleted.
(JSC::VMEntryScope::addDidPopListener): Deleted.
(JSC::VMEntryScope::~VMEntryScope): Deleted.
* Source/JavaScriptCore/runtime/VMEntryScope.h:
* Source/JavaScriptCore/runtime/VMEntryScopeInlines.h: Copied from Source/JavaScriptCore/runtime/VMEntryScope.h.
(JSC::VMEntryScope::VMEntryScope):
(JSC::VMEntryScope::~VMEntryScope):
* Source/JavaScriptCore/runtime/VMTraps.cpp:
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
* Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
* Source/WebCore/bindings/js/JSErrorHandler.cpp:
* Source/WebCore/bindings/js/JSEventListener.cpp:

Canonical link: <a href="https://commits.webkit.org/260858@main">https://commits.webkit.org/260858@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/907aa0ad8a6bdbd7bb3b47c95982e028c2a565d8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109707 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18824 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/42448 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/1182 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118827 "Built successfully") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/11/builds/113655 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/20294 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/10023 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101974 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115457 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/15105 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/98337 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43327 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/97078 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/29988 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/85113 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/1/builds/98577 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11553 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/31330 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/99454 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/9584 "Built successfully and passed tests") | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/82/builds/12205 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/8266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/46/builds/31109 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/17574 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/50939 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/107469 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13953 "Built successfully") | | [  ~~🧪 jsc-mips-tests~~](https://ews-build.webkit.org/#/builders/45/builds/26523 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4089 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian commented on Source/JavaScriptCore/interpreter/Interpreter.cpp:
```diff
@@ -1005,8 +1003,11 @@ JSValue Interpreter::executeProgram(const SourceCode& source, JSGlobalObject*, J
+            if (UNLIKELY(vm.traps().needHandling(VMTraps::AllEvents))) {
+                if (vm.hasExceptionsAfterHandlingTraps())
+                    return throwScope.exception();
+            }
```
This is wrong.  This bypasses the DeferTraps scope.  Instead, let's introduce a new macro as follows and use that:
```
#define RETURN_IF_EXCEPTION_WITH_TRAPS_DEFERRED(scope__, value__) do { \
        JSC::VM& vm = (scope__).vm(); \
        if (UNLIKELY(vm.exception())) \
            return value__; \
    } while (false)
```

Constellation replied:
Fixed.

MenloDorian reviewed (approved):
r=me with the RETURN_IF_EXCEPTION_WITH_TRAPS_DEFERRED fix.

Constellation committed:
[JSC] Make C++ -> JS calls faster
https://bugs.webkit.org/show_bug.cgi?id=252961
rdar://105943509

Reviewed by Mark Lam.

1. VMEntryScope's fast path should be inlined. Do it to make it work when it is used outside of JSC framework (WebCore etc.).
2. Move VMEntryScope's didPopListener to VM so that it makes VMEntryScope's destructor super simple (purge Vector destructor code bloat).
3. Remove isCollectorBusyOnCurrentThread check. It is extremely costly for this function (5% is attributed to this!). It is too costly
   for runtime check. And we instead have ASSERT already, which should catch an issue from fuzzers (And so far, we are not seeing this).
4. RETURN_IF_EXCEPTION inside DeferTraps always get the slow path function call! We call `vm.traps().needHandling(VMTraps::AllEvents)` manually
   to avoid this.

This patch makes C++ -> JS calls 8% faster.

                               ToT                     Patched

    cpp-to-js-call       85.4401+-0.4316     ^     78.9938+-0.3521        ^ definitely 1.0816x faster

* JSTests/microbenchmarks/cpp-to-js-call.js: Added.
(test):
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/debugger/Debugger.cpp:
* Source/JavaScriptCore/dfg/DFGOperations.cpp:
* Source/JavaScriptCore/interpreter/CallFrame.cpp:
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::executeProgram):
(JSC::Interpreter::executeCallImpl):
(JSC::Interpreter::executeConstruct):
(JSC::Interpreter::prepareForRepeatCall):
(JSC::Interpreter::executeEval):
(JSC::Interpreter::executeModuleProgram):
* Source/JavaScriptCore/runtime/StringPrototype.cpp:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::whenIdle):
* Source/JavaScriptCore/runtime/VM.h:
(JSC::VM::invokeDidPopListeners):
* Source/JavaScriptCore/runtime/VMEntryScope.cpp:
(JSC::VMEntryScope::setUpSlow):
(JSC::VMEntryScope::tearDownSlow):
(JSC::VMEntryScope::VMEntryScope): Deleted.
(JSC::VMEntryScope::addDidPopListener): Deleted.
(JSC::VMEntryScope::~VMEntryScope): Deleted.
* Source/JavaScriptCore/runtime/VMEntryScope.h:
* Source/JavaScriptCore/runtime/VMEntryScopeInlines.h: Copied from Source/JavaScriptCore/runtime/VMEntryScope.h.
(JSC::VMEntryScope::VMEntryScope):
(JSC::VMEntryScope::~VMEntryScope):
* Source/JavaScriptCore/runtime/VMTraps.cpp:
* Source/JavaScriptCore/tools/JSDollarVM.cpp:
* Source/WebCore/bindings/js/JSDOMGlobalObject.cpp:
* Source/WebCore/bindings/js/JSErrorHandler.cpp:
* Source/WebCore/bindings/js/JSEventListener.cpp:

Canonical link: https://commits.webkit.org/260858@main
